portality/bll/services/site.py

```diff
@@ -1,13 +1,57 @@
-from portality.core import app
-from portality import models
-from portality.store import StoreFactory, prune_container
+import re
 from datetime import datetime
-from portality.lib.argvalidate import argvalidate
+from typing import Iterable
+
+import werkzeug.routing
+from lxml import etree
 
+from portality import models
 from portality.bll import exceptions
+from portality.core import app
+from portality.lib.argvalidate import argvalidate
+from portality.store import StoreFactory, prune_container
+from portality.util import get_full_url_by_endpoint
+
+NS = "{http://www.sitemaps.org/schemas/sitemap/0.9}"
+
+
+def create_simple_sub_element(parent, element_name, text=None):
+    loc = etree.SubElement(parent, NS + element_name)
+    if text is not None:
+        loc.text = text
+    return loc
+
+
+def create_url_element(parent, loc, change_freq, lastmod=None):
+    url_ele = etree.SubElement(parent, NS + "url")
+
+    create_simple_sub_element(url_ele, 'loc', loc)
+    if lastmod is not None:
+        create_simple_sub_element(url_ele, "lastmod", lastmod)
+    create_simple_sub_element(url_ele, "changefreq", change_freq)
+
+    return url_ele
+
+
+def yield_all_route(entries: Iterable[dict]):
+    for e in entries:
+        if 'route' in e:
+            yield e['route']
+        if 'entries' in e:
+            yield from yield_all_route(e['entries'])
+
+
+def get_url_safe(endpoint):
```

Comment:
Propose to move this function to `portality.util` alongside our version of `url_for`